src/app/shell/item-comparators.ts

```diff
@@ -240,3 +349,69 @@ export function sortItems(
   );
   return items.toSorted(comparator);
 }
+
+export function groupItems(
+  items: readonly DimItem[],
+  vaultGrouping: string,
+  getTag: (item: DimItem) => TagValue | undefined,
+): readonly VaultGroup[] {
+  const groupingValueGetter = GROUPING_VALUE_GETTERS[vaultGrouping];
+  const iconValueGetter = ICON_VALUE_GETTERS[vaultGrouping];
+  const comparator = GROUP_BY_COMPARATORS[vaultGrouping];
+
+  // If there are no items, or the grouping is not suppored, return all items in a single group
+  if (!items.length || !groupingValueGetter || !comparator) {
+    return [{ groupingValue: undefined, items }];
+  }
+
+  const grouped: MutableVaultGroup[] = [];
+
+  for (const item of items) {
+    const indexOfUngrouped = grouped.findIndex((g) => g.groupingValue === undefined);
+
+    if (!groupingValueGetter) {
+      // If there is not an ungrouped group, create one
+      if (indexOfUngrouped < 0) {
+        grouped.push({ groupingValue: undefined, items: [item] });
+        continue;
+      }
+
+      // Add to existing ungrouped group
+      grouped[indexOfUngrouped].items.push(item);
+      continue;
+    }
+
+    const groupingValue = groupingValueGetter(item, getTag);
+
+    if (typeof groupingValue === 'undefined') {
```

Comment:
It seems sufficient to do this, like we do elsewhere:
```suggestion
    if (groupingValue === undefined) {
```